README.md
[![Latest Version](https://img.shields.io/github/release/juiceland/bbcodeparser.svg?style=flat-square)](https://github.com/juiceland/bbcodeparser/releases)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)
[![Build Status](https://img.shields.io/travis/juiceland/BBCodeParser/master.svg?style=flat-square)](https://travis-ci.org/juiceland/BBCodeParser)
[![Coverage Status](https://img.shields.io/scrutinizer/coverage/g/juiceland/bbcodeparser/master.svg?style=flat-square)](https://scrutinizer-ci.com/g/juiceland/bbcodeparser/code-structure)
[![Quality Score](https://img.shields.io/scrutinizer/g/juiceland/bbcodeparser/master.svg?style=flat-square)](https://scrutinizer-ci.com/g/juiceland/bbcodeparser)
[![Total Downloads](https://img.shields.io/packagist/dt/juiceland/bbcodeparser.svg?style=flat-square)](https://packagist.org/packages/juiceland/bbcodeparser)

The ``Golonka\BBCodeParser`` package will help you with parsing BBCode.

## Install

Via Composer

``` bash
$ composer require golonka/bbcodeparser
```

## Usage
To parse some text it's as easy as this!
``` php
$bbcode = new Golonka\BBCode\BBCodeParser;

echo $bbcode->parse('[b]Bold Text![/b]');
// <strong>Bold Text!</strong>
```
Would like the parser to not use all bbcodes? Just do like this.
``` php
$bbcode = new Golonka\BBCode\BBCodeParser;

echo $bbcode->only('bold', 'italic')
            ->parse('[b][u]Bold[/u] [i]Italic[/i]![/b]');
            // <strong>[u]Bold[/u] <em>Italic</em>!</strong>

echo $bbcode->except('bold')
            ->parse('[b]Bold[/b] [i]Italic[/i]');
            // [b]Bold[/b] <em>Italic</em>
```

By default the parser is case sensitive. But if you would like the parser to accept tags like `` [B]Bold Text[/B] `` it's really easy.
``` php
$bbcode = new Golonka\BBCode\BBCodeParser;

// Case insensitive
echo $bbcode->parse('[b]Bold[/b] [I]Italic![/I]', true);
     // <strong>Bold</strong> <em>Italic!</em>

// Or like this

echo $bbcode->parseCaseInsensitive('[b]Bold[/b] [i]Italic[/i]');
     // <strong>Bold</strong> <em>Italic!</em>
```
You could also make it more explicit that the parser is case sensitive by using another helper function.
``` php
    $bbcode = new Golonka\BBCode\BBCodeParser;

    echo $bbcode->parseCaseSensitive('[b]Bold[/b] [I]Italic![/I]');
         // <strong>Bold</strong> [I]Italic![/I]
```

If you would like to completely remove all BBCode it's just one function call away.
``` php
    $bbcode = new Golonka\BBCode\BBCodeParser;

    echo $bbcode->stripBBCodeTags('[b]Bold[/b] [i]Italic![/i]');
         // Bold Italic!
```

## Laravel integration
The integration into Laravel is really easy, and the method is the same for both Laravel 4 and Laravel 5.
Just open your ``app.php`` config file.

In there you just add this to your providers array
``` php
'Golonka\BBCode\BBCodeParserServiceProvider'
```

And this to your facades array
``` php
'BBCode' => 'Golonka\BBCode\Facades\BBCodeParser'
```

The syntax is the same as if you would use it in vanilla PHP but with the ``BBCode::`` before the methods.
Here are some examples.
``` php
// Simple parsing
echo BBCode::parse('[b]Bold Text![/b]');

// Limiting the parsers with the only method
echo BBCode::only('bold', 'italic')
        ->parse('[b][u]Bold[/u] [i]Italic[/i]![/b]');
        // <strong>[u]Bold[/u] <em>Italic</em>!</strong>

// Or the except method
echo BBCode::except('bold')
        ->parse('[b]Bold[/b] [i]Italic[/i]');
        // [b]Bold[/b] <em>Italic</em>
```

## Testing

``` bash
$ phpunit
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

## Credits

- [Joseph Landberg](https://github.com/golonka)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All Notable changes to `juiceland\BBCodeParser` will be documented in this file.

## v3.0.0 - 2019-07-30

Identical functionality compared to v2.3 but skipped support for PHP versions lower than v7.1.

- Bumped minimum PHP version to v7.1.
- Bumped phpunit version to ^7.5.

## v2.3.0 - 2019-07-30

This package was deleted from github some time ago because i couldn't find time to maintain it. Brought it back after some emails from different people that still used it. That motivated me to at least try and maintain it again and I will try to update some things over time. Sorry for the inconvenience, won't happen again!

### Added
- Added BBCode for sup, sub and small tags.

## v2.2.0 - 2015-09-07

### Added
- You can now strip all BBCode tags by using the ``stripBBCodeTags`` function.

## v2.1.0 - 2015-06-20

### Added
- Made ``parseCaseSensitive`` and ``parseCaseInsensitive`` functions to make parsing more readable.

## v2.0.0 - 2015-06-02

### Added
- Using PSR-4 instead of PSR-0
- Moved the ``arrayOnly`` and ``arrayExcept`` functions into a trait
- Minimum supported PHP version bumped to 5.4

### Fixed
- Renamed some tag names, mostly making them all lowercase
 - `` underLine -> underline ``
 - `` lineThrough -> linethrough ``
 - `` fontSize -> size ``
 - `` fontColor -> color ``
 - `` namedQuote -> namedquote ``
 - `` namedQuote -> namedquote ``
 - `` namedLink -> namedlink ``
 - `` orderedListNumerical -> orderedlistnumerical ``
 - `` orderedListAlpha -> orderedlistalpha ``
 - `` unorderedList -> unorderedlist ``
 - `` listItem -> listitem ``

### Removed
- The ``iterate`` property is removed. Unneeded after improvements in parsing method.
- Removed deprecated tags ``[ul]`` and ``[ol]``

## v1.4 - 2015-05-05

### Added
- Optional parameter enables or disables case insensitivity. Disabled by default.

## v1.3.0 - 2014-06-30

### Fixed
- The only/except functionally have been broken since like 1.1, but now it´s working. Better late then never!

## v1.2.7 - 2014-05-19

### Added
- A new iterate key is added to tags that typically could contain more tags of the same kind, like quotes.

### Fixed
- Problem where tags of the same kind would just parse the top level.

## v1.2.6 - 2014-05-17

### Fixed
- Fixed a problem where if a tag had a line break in them they wouldn't parse.

## v1.2.5 - 2014-05-15

### Fixed
- Improved most regex matches by removing unnecessary greediness.

## v1.2.0 - 2014-03-25

### Fixed
- Better syntax for lists.

## v1.1.0 - 2014-01-27

### Added
- Support for custom bbcode tags.

## v1.0.0 - 2013-11-07

Released the package.


tests/BBCodeParserTest.php
<?php

use \Golonka\BBCode\BBCodeParser;
use PHPUnit\Framework\TestCase;

class BBCodeParserTest extends TestCase {

    public function testBBCodeParserCanBeCreated()
    {
        $b = new BBCodeParser;
        $this->assertNotEmpty($b);
    }

    public function testSingleParsing()
    {
        $tests = array(
            array('in' => 'foo[b]bar[/b]baz', 'expected' => 'foo<strong>bar</strong>baz'),
            array('in' => 'foo[i]bar[/i]baz', 'expected' => 'foo<em>bar</em>baz'),
            array('in' => 'foo[s]bar[/s]baz', 'expected' => 'foo<strike>bar</strike>baz'),
            array('in' => 'foo[u]bar[/u]baz', 'expected' => 'foo<u>bar</u>baz'),
            array('in' => 'foo[size=6]bar[/size]baz', 'expected' => 'foo<font size="6">bar</font>baz'),
            array('in' => 'foo[color=#ff0000]bar[/color]baz', 'expected' => 'foo<font color="#ff0000">bar</font>baz'),
            array('in' => 'foo[color=#eee]bar[/color]baz', 'expected' => 'foo<font color="#eee">bar</font>baz'),
            array('in' => '[center]foobar[/center]', 'expected' => '<div style="text-align:center;">foobar</div>'),
            array('in' => '[left]foobar[/left]', 'expected' => '<div style="text-align:left;">foobar</div>'),
            array('in' => '[right]foobar[/right]', 'expected' => '<div style="text-align:right;">foobar</div>'),
            array('in' => '[quote]foobar[/quote]', 'expected' => '<blockquote>foobar</blockquote>'),
            array('in' => '[quote=golonka]foobar[/quote]', 'expected' => '<blockquote><small>golonka</small>foobar</blockquote>'),
            array('in' => '[url]http://www.aftonbladet.se[/url]', 'expected' => '<a href="http://www.aftonbladet.se">http://www.aftonbladet.se</a>'),
            array('in' => '[url=http://www.example.com]aftonbladet[/url]', 'expected' => '<a href="http://www.example.com">aftonbladet</a>'),
            array('in' => '[img]http://example.com/images/logo.png[/img]', 'expected' => '<img src="http://example.com/images/logo.png">'),
            array('in' => '[list=1][/list]', 'expected' => '<ol></ol>'),
            array('in' => '[list=a][/list]', 'expected' => '<ol type="a"></ol>'),
            array('in' => '[list][/list]', 'expected' => '<ul></ul>'),
            array('in' => '[*]Item 1', 'expected' => '<li>Item 1</li>'),
            array('in' => '[code]<?php echo \'Hello World\'; ?>[/code]', 'expected' => '<code><?php echo \'Hello World\'; ?></code>'),
            array('in' => '[youtube]Nizq4RnsJJo[/youtube]', 'expected' => '<iframe width="560" height="315" src="//www.youtube.com/embed/Nizq4RnsJJo" frameborder="0" allowfullscreen></iframe>'),
            array('in' => '[sub]Subscript[/sub]', 'expected' => '<sub>Subscript</sub>'),
            array('in' => '[sup]Superscript[/sup]', 'expected' => '<sup>Superscript</sup>'),
            array('in' => '[small]Small text[/small]', 'expected' => '<small>Small text</small>'),
        );
        $b = new BBCodeParser;

        foreach ($tests as $test) {
            $result = $b->parse($test['in']);
            $this->assertEquals($result, $test['expected']);
        }
    }

    public function testCompleteBBCodeParser()
    {
        $b = new BBCodeParser;
        $s = '
            [b]bold[/b][i]italic[/i][u]underline[/u][s]line through[/s][size=6]size[/size]
            [color=#eee]color[/color][center]centered text[/center][quote]quote[/quote]
            [quote=golonka]quote[/quote][url]http://www.example.com[/url]
            [url=http://www.example.com]example.com[/url][img]http://example.com/logo.png[/img]
            [list=1]
                [*]Item 1
                [*]Item 2
                [*]Item 3
            [/list]
            [code]<?php echo \'Hello World\'; ?>[/code]
            [youtube]Nizq4RnsJJo[/youtube]
            [list]
                [*]Item 1
                [*]Item 2
                [*]Item 3
            [/list]
        ';
        $r = $b->parse($s);
        $this->assertEquals('
            <strong>bold</strong><em>italic</em><u>underline</u><strike>line through</strike><font size="6">size</font>
            <font color="#eee">color</font><div style="text-align:center;">centered text</div><blockquote>quote</blockquote>
            <blockquote><small>golonka</small>quote</blockquote><a href="http://www.example.com">http://www.example.com</a>
            <a href="http://www.example.com">example.com</a><img src="http://example.com/logo.png">
            <ol>
                <li>Item 1</li>
                <li>Item 2</li>
                <li>Item 3</li>
            </ol>
            <code><?php echo \'Hello World\'; ?></code>
            <iframe width="560" height="315" src="//www.youtube.com/embed/Nizq4RnsJJo" frameborder="0" allowfullscreen></iframe>
            <ul>
                <li>Item 1</li>
                <li>Item 2</li>
                <li>Item 3</li>
            </ul>
        ', $r);
    }

    public function testOnlyFunctionality()
    {
        $b = new BBCodeParser;

        $b->only('bold', 'underline');
        $this->arrays_are_similar($b->getParsers(), ['image', 'link']);

        $result = $b->parse('[b]Bold[/b] [url]http://example.com[/url] [u]Underline[/u]');
        $this->assertEquals($result, '<strong>Bold</strong> [url]http://example.com[/url] <u>Underline</u>');
    }

    public function testExceptFunctionality()
    {
        $b = new BBCodeParser;

        $b->except('link', 'bold');
        $this->arrays_are_similar(
            $b->getParsers(),
            [
                'italic',
                'underLine',
                'linethrough',
                'color',
                'center',
                'quote',
                'quote',
                'namedlink',
                'orderedlistnumerical',
                'orderedlistalpha',
                'unorderedlist',
                'listitem',
                'code',
                'youtube',
                'linebreak',
            ]
        );

        $result = $b->parse('[b]Bold[/b] [url]http://example.com[/url] [u]Underline[/u]');
        $this->assertEquals($result, '[b]Bold[/b] [url]http://example.com[/url] <u>Underline</u>');
    }

    public function testCustomParser()
    {
        $b = new BBCodeParser;

        $b->setParser('verybold', '/\[verybold\](.*)\[\/verybold\]/', '<strong>VERY $1 BOLD</strong>', '$1');

        $result = $b->parse('[verybold]something[/verybold]');

        $this->assertEquals($result, '<strong>VERY something BOLD</strong>');
    }

    public function testIfTagsGetParsedProperlyIfLineBreaksExists()
    {
        $b = new BBCodeParser;
        $result = $b->parse(
            't[b]e[/b]s[b]t[/b]
            [code]Test
            123[/code]'
        );

        $this->assertEquals(
            $result,
            't<strong>e</strong>s<strong>t</strong>
            <code>Test
            123</code>'
        );
    }

    public function testNestedNamedQuotes()
    {
        $b = new BBCodeParser;
        $result = $b->parse('[quote=Xzibit][quote=PimpMyRide]So i put a quote in you quote![/quote]I heard you liked quotes.[/quote]');

        $this->assertEquals(
            $result,
            '<blockquote><small>Xzibit</small><blockquote><small>PimpMyRide</small>So i put a quote in you quote!</blockquote>I heard you liked quotes.</blockquote>'
        );
    }

    public function testNestedQuotes()
    {
        $b = new BBCodeParser;
        $result = $b->parse('[quote][quote]Inception[/quote]Quoteception[/quote]');

        $this->assertEquals(
            $result,
            '<blockquote><blockquote>Inception</blockquote>Quoteception</blockquote>'
        );
    }

    public function testCaseSensitivity()
    {
        $b = new BBCodeParser;
        $result = $b->parse('[B][I][U]More tags === More COOL![/U][/I][/B]', true);

        $this->assertEquals(
            $result,
            '<strong><em><u>More tags === More COOL!</u></em></strong>'
        );
    }

    public function testCaseSensitivityHelperFunctions()
    {
        $b = new BBCodeParser;
        $result = $b->parseCaseSensitive('[b]Bold[/b] [I]Italic[/I]');
        $this->assertEquals($result, '<strong>Bold</strong> [I]Italic[/I]');

        $b = new BBCodeParser;
        $result = $b->parseCaseInsensitive('[b]Bold[/b] [I]Italic[/I]');
        $this->assertEquals($result, '<strong>Bold</strong> <em>Italic</em>');
    }

    public function testRemovalOfBBCodeTags()
    {
        $b = new BBCodeParser;
        $result = $b->stripBBCodeTags('[u]Magpie[/u] is one [b]beautiful[/b] bird!');
        $this->assertEquals($result, 'Magpie is one beautiful bird!');

        $b = new BBCodeParser;
        $result = $b->stripBBCodeTags('[QUOTE=Magpie]I am one [B]beautiful[/B] bird![/QUOTE]');
        $this->assertEquals($result, 'I am one beautiful bird!');
    }

    public function testStatelessParsers()
    {
        $b = new BBCodeParser;
        $result = $b->only('bold')->parse('[b]bold[/b] [i]italic[/i]');
        $this->assertEquals($result, '<strong>bold</strong> [i]italic[/i]');

        $result = $b->only('italic')->parse('[b]bold[/b] [i]italic[/i]');
        $this->assertEquals($result, '[b]bold[/b] <em>italic</em>');
    }

    protected function arrays_are_similar($a, $b) {
        // if the indexes don't match, return immediately
        if (count(array_diff_assoc($a, $b))) {
            return false;
        }
        // we know that the indexes, but maybe not values, match.
        // compare the values between the two arrays
        foreach($a as $k => $v) {
            if ($v !== $b[$k]) {
                return false;
            }
        }
        // we have identical indexes, and no unequal values
        return true;
    }

}


src/BBCodeParser.php
<?php namespace Golonka\BBCode;

use \Golonka\BBCode\Traits\ArrayTrait;

class BBCodeParser
{

    use ArrayTrait;

    public $parsers = [
        'bold' => [
            'pattern' => '/\[b\](.*?)\[\/b\]/s',
            'replace' => '<strong>$1</strong>',
        ],
        'italic' => [
            'pattern' => '/\[i\](.*?)\[\/i\]/s',
            'replace' => '<em>$1</em>',
        ],
        'underline' => [
            'pattern' => '/\[u\](.*?)\[\/u\]/s',
            'replace' => '<u>$1</u>',
        ],
        'linethrough' => [
            'pattern' => '/\[s\](.*?)\[\/s\]/s',
            'replace' => '<strike>$1</strike>',
        ],
        'size' => [
            'pattern' => '/\[size\=([1-7])\](.*?)\[\/size\]/s',
            'replace' => '<font size="$1">$2</font>',
        ],
        'color' => [
            'pattern' => '/\[color\=(#[A-f0-9]{6}|#[A-f0-9]{3})\](.*?)\[\/color\]/s',
            'replace' => '<font color="$1">$2</font>',
        ],
        'center' => [
            'pattern' => '/\[center\](.*?)\[\/center\]/s',
            'replace' => '<div style="text-align:center;">$1</div>',
        ],
        'left' => [
            'pattern' => '/\[left\](.*?)\[\/left\]/s',
            'replace' => '<div style="text-align:left;">$1</div>',
        ],
        'right' => [
            'pattern' => '/\[right\](.*?)\[\/right\]/s',
            'replace' => '<div style="text-align:right;">$1</div>',
        ],
        'quote' => [
            'pattern' => '/\[quote\](.*?)\[\/quote\]/s',
            'replace' => '<blockquote>$1</blockquote>',
        ],
        'namedquote' => [
            'pattern' => '/\[quote\=(.*?)\](.*)\[\/quote\]/s',
            'replace' => '<blockquote><small>$1</small>$2</blockquote>',
        ],
        'link' => [
            'pattern' => '/\[url\](.*?)\[\/url\]/s',
            'replace' => '<a href="$1">$1</a>',
        ],
        'namedlink' => [
            'pattern' => '/\[url\=(.*?)\](.*?)\[\/url\]/s',
            'replace' => '<a href="$1">$2</a>',
        ],
        'image' => [
            'pattern' => '/\[img\](.*?)\[\/img\]/s',
            'replace' => '<img src="$1">',
        ],
        'orderedlistnumerical' => [
            'pattern' => '/\[list=1\](.*?)\[\/list\]/s',
            'replace' => '<ol>$1</ol>',
        ],
        'orderedlistalpha' => [
            'pattern' => '/\[list=a\](.*?)\[\/list\]/s',
            'replace' => '<ol type="a">$1</ol>',
        ],
        'unorderedlist' => [
            'pattern' => '/\[list\](.*?)\[\/list\]/s',
            'replace' => '<ul>$1</ul>',
        ],
        'listitem' => [
            'pattern' => '/\[\*\](.*)/',
            'replace' => '<li>$1</li>',
        ],
        'code' => [
            'pattern' => '/\[code\](.*?)\[\/code\]/s',
            'replace' => '<code>$1</code>',
        ],
        'youtube' => [
            'pattern' => '/\[youtube\](.*?)\[\/youtube\]/s',
            'replace' => '<iframe width="560" height="315" src="//www.youtube.com/embed/$1" frameborder="0" allowfullscreen></iframe>',
        ],
        'linebreak' => [
            'pattern' => '/\r/',
            'replace' => '<br />',
        ]
    ];

    /**
     * Parses the BBCode string
     * @param  string $source String containing the BBCode
     * @return string Parsed string
     */
    public function parse($source, $caseInsensitive = false)
    {
        foreach ($this->parsers as $name => $parser) {
            $pattern = ($caseInsensitive) ? $parser['pattern'].'i' : $parser['pattern'];

            while (preg_match($pattern, $source)) {
                $source = preg_replace($pattern, $parser['replace'], $source);
            }
        }
        return $source;
    }

    /**
     * Helper function to parse case sensitive
     * @param  string $source String containing the BBCode
     * @return string Parsed text
     */
    public function parseCaseSensitive($source)
    {
        return $this->parse($source, false);
    }

    /**
     * Helper function to parse case insensitive
     * @param  string $source String containing the BBCode
     * @return string Parsed text
     */
    public function parseCaseInsensitive($source)
    {
        return $this->parse($source, true);
    }

    /**
     * Limits the parsers to only those you specify
     * @param  mixed $only parsers
     * @return object BBCodeParser object
     */
    public function only($only = null)
    {
        $only = (is_array($only)) ? $only : func_get_args();
        $this->parsers = $this->arrayOnly($this->parsers, $only);
        return $this;
    }

    /**
     * Removes the parsers you want to exclude
     * @param  mixed $except parsers
     * @return object BBCodeParser object
     */
    public function except($except = null)
    {
        $except = (is_array($except)) ? $except : func_get_args();
        $this->parsers = $this->arrayExcept($this->parsers, $except);
        return $this;
    }

    /**
     * List of chosen parsers
     * @return array array of parsers
     */
    public function getParsers()
    {
        return $this->parsers;
    }

    /**
     * Sets the parser pattern and replace.
     * This can be used for new parsers or overwriting existing ones.
     * @param string $name Parser name
     * @param string $pattern Pattern
     * @param string $replace Replace pattern
     * @return void
     */
    public function setParser($name, $pattern, $replace)
    {
        $this->parsers[$name] = array(
            'pattern' => $pattern,
            'replace' => $replace
        );
    }
}
